### PR TITLE
Add fair distribution options

### PR DESCRIPTION
**Pull request by SouthEndMusic**
Fixes https://github.com/Deltares/Ribasim/issues/726.

### CODE REVIEW

**Comment by SouthEndMusic** [comment]
To do: remove demands as capacity constraints, are now in objective function.

**Comment by SouthEndMusic** [comment]
As a follow-up to https://github.com/Deltares/Ribasim/pull/743#issuecomment-1798465701, here are some allocation results that are more fair:

simple_subnetwork
allocation_result_quadratic_absolute
allocation_result_quadratic_relative

**Comment by gijsber** [comment]
Nice :-). in quadratic_relative, is 5 going up or 6 going down?

File: core/src/allocation.jl [hunk 1]
```diff
@@ -386,35 +386,34 @@ function add_variables_flow!(
...
+# Non-negativivity constraints are also immediately added to the basin allocation variables.
+# """
+# function add_variables_allocation_basin!(
+#     problem::JuMP.Model,
+#     allocgraph_node_ids_basin::Vector{Int},
+# )::Nothing
+#     JuMP.@variable(problem, A_basin[i = allocgraph_node_ids_basin] >= 0.0)
+#     return nothing
+# end
+
```
Comment (visr) [inline core/src/allocation.jl:401]: If this isn't used anymore I'd prefer to remove it, we can always take it out of git history if needed.

File: core/src/allocation.jl [hunk 2]
```diff
@@ -591,11 +626,11 @@ function allocation_problem(
...
     # Add variables to problem
     add_variables_flow!(problem, allocgraph_edges)
-    add_variables_allocation_basin!(problem, node_id_mapping, allocgraph_node_ids_basin)
+    # add_variables_allocation_basin!(problem, allocgraph_node_ids_basin)
+    add_variables_absolute_value!(problem, allocgraph_edge_ids_user_demand, config)
 
     # Add constraints to problem
-    add_constraints_user_allocation!(problem, allocgraph_edge_ids_user_demand)
-    add_constraints_basin_allocation!(problem, allocgraph_node_ids_basin)
+    # add_constraints_basin_allocation!(problem, allocgraph_node_ids_basin)
```
Comment (visr) [inline core/src/allocation.jl:633]: Same.

File: core/src/allocation.jl [hunk 3]
```diff
@@ -610,10 +645,11 @@ function allocation_problem(
         allocgraph_node_outedge_ids,
     )
     add_constraints_user_returnflow!(problem, allocgraph_node_ids_user_with_returnflow)
+    add_constraints_absolute_value!(problem, allocgraph_edge_ids_user_demand, config)
     # TODO: The fractional flow constraints
 
-    # Add objective to problem
-    add_objective_function!(problem, allocgraph_edge_ids_user_demand)
+    # # Add objective to problem
+    # add_objective_function!(problem, allocgraph_edge_ids_user_demand, config)
```
Comment (visr) [inline core/src/allocation.jl:652]: Same.

File: core/src/allocation.jl [hunk 4]
```diff
@@ -818,7 +899,12 @@ function allocate!(p::Parameters, allocation_model::AllocationModel, t::Float64)
         # or set edge capacities if priority_idx = 1
         adjust_edge_capacities!(allocation_model, priority_idx)
 
-        set_demands_priority!(allocation_model, user, priority_idx, t)
+        # Set the objective depending on the demands
+        # A new objective function is set instead of modifying the coefficients
+        # of an existing objective function because this is not supported for
+        # quadratic terms:
+        # https://jump.dev/JuMP.jl/stable/manual/objective/#Modify-an-objective-coefficient
```
Comment (visr) [inline core/src/allocation.jl:906]:
```suggestion
        # https://jump.dev/JuMP.jl/v1.16/manual/objective/#Modify-an-objective-coefficient
```

Link is less likely to die this way.

File: docs/core/allocation.qmd [hunk 5]
```diff
@@ -94,25 +94,39 @@ There are 2 types of variables whose value has to be determined to solve the all
...
+where $\hat{B}_S = m_S(B_S) \subset \hat{V}_S$ is the set of basin node ids in the allocation graph.
 
 :::{.callout-note}
 Currently the basin allocations are not taken into account in the implementation.
 :::
 
 ### The optimization objective
 
-The goal of allocation is to maximize the flow to the users. Therefore, we use the following optimization objective:
+The goal of allocation is to get the flow to the users as close as possible to their demand. To achieve this, the following objectives are supported:
```
Comment (visr) [inline docs/core/allocation.qmd:105]: Besides the equations, could you add a bit more text here on the differences? Like that quadratic focuses more on avoiding large shortages, and that for linear it is all the same.

**Review by visr — APPROVED** [review]

**Pull request merged**